File: scripts/governance/check-silent-catches.js
#!/usr/bin/env node
/**
 * Iron Dome Ratchet — silent/empty catch blocks
 * Blocks: } catch (e) {} or } catch { // nothing }
 */

const fs = require('fs');
const path = require('path');

const BASELINE_FILE = '.memory-layer/baselines/type-safety.json';

function countSilentCatches(dir) {
  let count = 0;
  function walk(d) {
    for (const f of fs.readdirSync(d)) {
      const full = path.join(d, f);
      const stat = fs.statSync(full);
      if (stat.isDirectory() && !['node_modules', 'dist', '.git', 'coverage'].includes(f)) {
        walk(full);
      } else if (/\.(ts|tsx|js|jsx)$/.test(f) && !f.endsWith('.d.ts')) {
        const content = fs.readFileSync(full, 'utf8');
        // catch block with only whitespace/comments
        const silentCatch = content.match(/catch\s*(?:\([^)]*\))?\s*\{\s*(?:\/\/[^\n]*)?\s*\}/g) || [];
        count += silentCatch.length;
      }
    }
  }
  walk(dir);
  return count;
}

if (!fs.existsSync(BASELINE_FILE)) {
  console.log('✓ Silent catches: no baseline yet');
  process.exit(0);
}

const baseline = JSON.parse(fs.readFileSync(BASELINE_FILE, 'utf8'));
const current = countSilentCatches('src');
const limit = baseline.silent_catches ?? 999;

if (current > limit) {
  console.error(`✗ Silent catches increased: ${limit} → ${current}. Add proper error handling.`);
  process.exit(1);
}

console.log(`✓ Silent catches: ${current}/${limit} (baseline holds)`);
process.exit(0);

File: scripts/governance/check_silent_catches.py
#!/usr/bin/env python3
"""
Iron Dome Ratchet — bare except: pass or except Exception: pass blocks
"""

import ast
import json
import sys
from pathlib import Path

BASELINE_FILE = Path(".memory-layer/baselines/type-safety.json")


def count_silent_catches(directory: Path) -> int:
    count = 0
    for f in directory.rglob("*.py"):
        if any(p in str(f) for p in ["__pycache__", ".venv", "venv"]):
            continue
        try:
            tree = ast.parse(f.read_text())
        except SyntaxError:
            continue
        for node in ast.walk(tree):
            if isinstance(node, ast.ExceptHandler):
                body = node.body
                if all(
                    isinstance(s, ast.Pass)
                    or (isinstance(s, ast.Expr) and isinstance(s.value, ast.Constant))
                    for s in body
                ):
                    count += 1
    return count


if not BASELINE_FILE.exists():
    print("✓ Silent catches: no baseline yet")
    sys.exit(0)

baseline = json.loads(BASELINE_FILE.read_text())
app_dir = Path("app")
if not app_dir.exists():
    print("✓ Silent catches: no app/ directory")
    sys.exit(0)

current = count_silent_catches(app_dir)
limit = baseline.get("python_silent_catches", 999)

if current > limit:
    print(f"✗ Silent catches increased: {limit} → {current}", file=sys.stderr)
    sys.exit(1)

print(f"✓ Silent catches: {current}/{limit} (baseline holds)")
